plutus-metatheory/src/Check.lagda.md

```diff
@@ -85,220 +81,303 @@ data TypeError : Set where
-    where A ⇒ B ,, _ → inj₁ (notPi (A ⇒ B) (λ _ ()))
-          ne A  ,, _ → inj₁ (notPi (ne A) (λ _ ()))
-          con {Φ} c ,, _ → inj₁ (notPi (con {Φ} c) (λ _ ()))
-          μ A B ,, _ → inj₁ (notPi (μ A B) (λ _ ()))
-  return (_ ,, A ,, L)
+isPi (Π A ,, L)       = return (_ ,, A ,, L)
+isPi (A ⇒ B ,, _)     = inj₁ (notPi (A ⇒ B) (λ _ ()))
+isPi (ne A  ,, _)     = inj₁ (notPi (ne A) (λ _ ()))
+isPi (con {Φ} c ,, _) = inj₁ (notPi (con {Φ} c) (λ _ ()))
+isPi (μ A B ,, _)     = inj₁ (notPi (μ A B) (λ _ ()))
+  
 
 isFunType :  ∀{Φ Γ}
-       → Either TypeError (Σ (Φ ⊢Nf⋆ *) (Γ ⊢_))
+       → Σ (Φ ⊢Nf⋆ *) (Γ ⊢_)
        → Either TypeError (Σ (Φ ⊢Nf⋆ *) λ A → Σ (Φ ⊢Nf⋆ *) λ B → Γ ⊢ A ⇒ B)
-isFunType p = do
-  A ⇒ B ,, L ← p
-    where Π A ,, _ → inj₁ (notFunType (Π A) (λ _ _ ()))
-          ne A  ,, _ → inj₁ (notFunType (ne A) (λ _ _ ()))
-          con {Φ} c ,, _ → inj₁ (notFunType (con {Φ} c) (λ _ _ ()))
-          μ A B ,, _ → inj₁ (notFunType (μ A B) (λ _ _ ()))
-  return (A ,, B ,, L)
+isFunType (A ⇒ B ,, L ) =  return (A ,, B ,, L)
+isFunType (Π A ,, _) = inj₁ (notFunType (Π A) (λ _ _ ()))
+isFunType (ne A  ,, _ ) = inj₁ (notFunType (ne A) (λ _ _ ()))
+isFunType (con {Φ} c ,, _) = inj₁ (notFunType (con {Φ} c) (λ _ _ ()))
+isFunType (μ A B ,, _) = inj₁ (notFunType (μ A B) (λ _ _ ()))
+ 
   
 isMu :  ∀{Φ Γ}
-       → Either TypeError (Σ (Φ ⊢Nf⋆ *) (Γ ⊢_))
+       → Σ (Φ ⊢Nf⋆ *) (Γ ⊢_)
        → Either TypeError (Σ Kind λ K → Σ (Φ ⊢Nf⋆ (K ⇒ *) ⇒ (K ⇒ *)) λ A → Σ (Φ ⊢Nf⋆ K) λ B → Γ ⊢ μ A B)
-isMu p = do
-  μ A B ,, L ← p
-    where Π A ,, _ → inj₁ (notMu (Π A) (λ _ _ ()))
-          ne A  ,, _ → inj₁ (notMu (ne A) (λ _ _ ()))
-          con {Φ} c ,, _ → inj₁ (notMu (con {Φ} c) (λ _ _ ()))
-          A ⇒ B ,, _ → inj₁ (notMu (A ⇒ B) (λ _ _ ()))
-  return (_ ,, A ,, B ,, L)
+isMu (μ A B ,, L) = return (_ ,, A ,, B ,, L)
+isMu (Π A ,, _) = inj₁ (notMu (Π A) (λ _ _ ()))
+isMu (ne A  ,, _) = inj₁ (notMu (ne A) (λ _ _ ()))
+isMu (con {Φ} c ,, _) = inj₁ (notMu (con {Φ} c) (λ _ _ ()))
+isMu (A ⇒ B ,, _) = inj₁ (notMu (A ⇒ B) (λ _ _ ()))
+```
+We have two base kinds, * and ♯, and we have an embedding of ♯ into *
+
+   con : Φ ⊢⋆ ♯
+         ------
+       → Φ ⊢⋆ *
 
+Hence, when inferring a kind we sometimes nned to decide if we want a ♯ kind or a * kind. For example,
+
+  con (atomic aInteger) : ScopedTy 0 
+
+  might be inferred as kind ♯
+
+  1.    ne (^ (atomic aInteger))
+  
+  or as kind *
+  
+  2.    con (ne (^ (atomic aInteger)))
+
+Whenever we have a constant of kind `♯` we embed it into `*` using `con`.
+This means that a composition of, for instance a list (kind `♯ ⇒ ♯`) applied 
+to a constant (which will be of kind `*`) doesn't match exactly. So we relax 
+this condition  when checking kinds and allow
+  1. Checking a `*`-kinded type against `♯`, whenever the type is of the form `con A` : 
+       we "downgrade" the type to A of kind `♯`. 
+  2. Checking a `♯`-kinded type against `*`: 
+       we "upgrade" the type to `*` using `con`.
```

Comment:
One problem is that a standalone `con (atomic aInteger)` should be inferred as a kind star * (i.e. 
`con (ne (^ (atomic aInteger)))`. I guess one could try to rewrite the inferKind function as a two-step process, but I'm not sure it would work in general. I'll try.

Reply:
Ok. Now I did implemented upgrading "as needed" and I remembered why we need to prematurely upgrade. You have functions like `isPat` that checks a type has kind `(K ⇒ *) ⇒ (K ⇒ *)` and returns a type `Φ ⊢Nf⋆ (K ⇒ *) ⇒ (K ⇒ *)`.  
If we have a kind, say `(K ⇒ ♯) ⇒ (K ⇒ ♯)` then we have the requirement to upgrade `♯`, but we can't get inside the type to insert the appropriate `con`s. (Well, we probably can make a function to do that, but that is even messier than the solution above.)

Reply:
I'll add an example for this case.